discv5/discv5-wire.md

```diff
@@ -49,187 +51,107 @@ request/response and 1s for the handshake.
+    id-nonce      = uint256   -- random bytes
+    enr-seq       = uint64    -- ENR sequence number of the requesting node
 
-    whoareyou-packet = magic || [token, id-nonce, enr-seq]
-    magic            = sha256(dest-node-id || "WHOAREYOU")
-    token            = auth-tag of request
-    id-nonce         = 32 random bytes
-    enr-seq          = highest ENR sequence number of node A known on node B's side
+![whoareyou packet layout](./img/whoareyou-packet-layout.png)
 
-The first encrypted message sent in response to WHOAREYOU contains an authentication
-header completing the handshake. The plain text of the authentication response is.
+### Handshake Message Packet (`flag = 2`)
 
-    auth-response-pt = [version, id-nonce-sig, node-record]
-    version          = 5
-    id-nonce-input   = sha256("discovery-id-nonce" || id-nonce || ephemeral-key)
-    id-nonce-sig     = id_sign(id-nonce-input)
-    static-node-key  = the private key used for node record identity
-    node-record      = record of sender OR [] if enr-seq in WHOAREYOU != current seq
+For handshake message packets, the `authdata` section has variable size since public key
+and signature sizes depend on the ENR identity scheme. For the "v4" identity scheme, we
+assume 64-byte signature size and 33 bytes of (compressed) public key size.
 
-`auth-response-pt` is encrypted with a separate key, the `auth-resp-key`, using an
-all-zero nonce. This is safe because only one message is ever encrypted with this key.
+`authdata` starts with a fixed-size `authdata-head` component, followed by the ID
+signature, ephemeral public key and optional node record.
 
-    message-packet   = tag || auth-header || message
-    auth-header      = [auth-tag, id-nonce, auth-scheme-name, ephemeral-pubkey, auth-response]
-    auth-scheme-name = "gcm"
-    auth-response    = aesgcm_encrypt(auth-resp-key, zero-nonce, auth-response-pt, "")
-    zero-nonce       = 12 zero bytes
-    message          = aesgcm_encrypt(initiator-key, auth-tag, message-pt, tag)
-    message-pt       = message-type || message-data
-    auth-tag         = AES-GCM nonce, 12 random bytes unique to message
+The `record` field may be omitted if the `enr-seq` of WHOAREYOU is recent enough, i.e.
+when it matches the current sequence number of the sending node. If `enr-seq` is zero, the
+record must be sent. Node records are encoded and verified as specified in [EIP-778].
 
-All messages following the handshake are encoded as follows:
+Please refer to the [handshake section] for more information about the content of the
+handshake packet.
 
-    message-packet   = tag || rlp_bytes(auth-tag) || message
-    message          = aesgcm_encrypt(initiator-key, auth-tag, message-pt, tag)
+    authdata      = authdata-head || id-signature || eph-pubkey || record
+    authdata-size = 15 + sig-size + eph-key-size + len(record)
+    authdata-head = version || nonce || sig-size || eph-key-size
+    version       = uint8     -- value: 1
```

Comment:
What does this version relate to?

Reply:
I see. What part of the wire protocol could be affected by changing this version?

Reply:
Anything after the handshake could change. We can add new messages, change existing messages, or change the handshake crypto.

Reply:
If it could affect almost everything, doesn't it make sense to add a version to a static header? We have a couple of spare bytes there.

Reply:
I don't want to put it into the static header because that would increase the size of all packets.

Reply:
> I don't want to put it into the static header because that would increase the size of all packets.

Don't we have a couple of unused bytes in `protocol-id` field, which are currently filled with spaces? 

Version does make sense for all packets until handshake is done: `RANDOM_PACKET`, `WHOAREYOU`, `HANDSHAKE`. As layout of each of these packets could be affected by changes in crypto. Having a version in all those packets, also, allows for backward compatibility and smoother upgradability.

Reply:
Yes, you're right, we could put it in this place.